### PR TITLE
Create DEPLOYING.md to document deployment instructions.

### PR DESCRIPTION
**Pull request by zhangkun83**
@nmittler please review.

### CODE REVIEW

**Comment by ejona86** [comment]
Should this go in a different file since most people shouldn't need to look at it and it could scare new users?

**Comment by zhangkun83** [comment]
Done. PTAL.

File: DEPLOYING.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+How to Deploy to Maven Central (for Maintainers Only)
+-----------------------------------------------------
+
+If you haven't deployed artifacts to Maven Central before:
+- Follow the instructions on [this
+  page](http://central.sonatype.org/pages/ossrh-guide.html) to set up an
+  account with OSSRH.
+- Install GnuPG and [generate your key pair](https://www.gnupg.org/documentation/howtos.html)
+- Put your GnuPG key password and OSSRH account information in
```
Comment (nmittler) [inline DEPLOYING.md]: You actually don't need GnuPG to make this work.  You can just use a username/password.

@ejona86  thoughts?

File: DEPLOYING.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+The Gradle task for deploying artifacts to Maven Central is ``uploadArchives``.
+However, because the codegen, which is a native executable, is built and
+deployed by this task too, you usually need to set additional environment
+variables before running this task. Please refer to _Pushing the codegen to
+Maven Central_ from ``compiler/README.md`` for how to set these environment
+variables on various platforms.
+
+Running ``uploadArchives`` from the top-level project will deploy the codegen
+for one or two platforms, depending on what you put in ``TARGET_ARCHS``. You
+need to run ``uploadArchives`` under ``compiler/`` on the rest of the supported
```
Comment (nmittler) [inline DEPLOYING.md]: The compiler instructions had the command that I could copy & paste ... could we just put that here instead?

File: DEPLOYING.md (hunk 1 above)
Comment (ejona86) [inline DEPLOYING.md]: You do but you don't. If making a snapshot you don't. If making a release you do. Maybe we can say:
"For release deployments, you will need to put your GnuPG..."

Comment (nmittler) [inline DEPLOYING.md]: sgtm

**Comment by nmittler** [comment]
LGTM

**Comment by zhangkun83** [comment]
@ejona86 I have overhauled the codegen compilation instructions. PTAL.

**Comment by ejona86** [comment]
@zhangkun83, LGTM. Very nice documentation.

**Pull request merged**